telemetry/ttl.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+const (
+	// selectDeletedRowsOneDaySQL selects the deleted rows for each table of last day
+	selectDeletedRowsOneDaySQL = `SELECT parent_table_id, CAST(SUM(deleted_rows) AS SIGNED)
+		FROM
+		    mysql.tidb_ttl_job_history
+		WHERE
+		    create_time >= CURDATE() - INTERVAL 7 DAY
+			AND finish_time >= CURDATE() - INTERVAL 1 DAY
+			AND finish_time < CURDATE()
+		GROUP BY parent_table_id;`
+	// selectDelaySQL selects the deletion delay in minute for each table at the end of last day
+	selectDelaySQL = `SELECT
```

Comment:
I still consider that we could get an accurate delay for each task during scan through:

```
MySQL [test]> explain select pk, max(created_at - NOW()) over() from t where created_at > NOW(
);
+--------------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------------+
| id                             | estRows  | task      | access object | operator info                                                                          |
+--------------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------------+
| Projection_8                   | 3333.33  | root      |               | test.t.pk, Column#6                                                                    |
| └─Window_9                     | 3333.33  | root      |               | max(Column#5)->Column#6 over()                                                         |
|   └─Projection_10              | 3333.33  | root      |               | test.t.pk, minus(cast(test.t.created_at, bigint(19) BINARY), 20230129144742)->Column#5 |
|     └─TableReader_14           | 3333.33  | root      |               | data:Selection_13                                                                      |
|       └─Selection_13           | 3333.33  | cop[tikv] |               | gt(test.t.created_at, 2023-01-29 14:47:42)                                             |
|         └─TableFullScan_12     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                                                         |
+--------------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------------+
6 rows in set (0.001 sec)
```

This window function should only scan once :thinking:. We could collect these information in the stats (that json column), and aggregate them in the history. How do you think about it?

But I agree that this PR is good enough, I'd try to make it better in the future  :+1:

Reply:
Oops, nothing. It only collects the newest expired row in the last scan, but not the oldest not-expired row in this scan.